assets/js/amp-block-validation.js

```diff
@@ -0,0 +1,177 @@
+				'amp/add-notice',
+				module.conditionallyAddNotice
+			);
+
+			module.store = wp.data.registerStore( 'amp/blockValidation', {
+				reducer: function( _state, action ) {
+					var state = _state || {
+						blockValidationErrorsByUid: {}
+					};
+
+					switch ( action.type ) {
+						case 'UPDATE_BLOCKS_VALIDATION_ERRORS':
+							return _.extend( {}, state, {
+								blockValidationErrorsByUid: action.blockValidationErrorsByUid
+							} );
+						default:
+							return state;
+					}
+				},
+				actions: {
+					updateBlocksValidationErrors: function( blockValidationErrorsByUid ) {
+						return {
+							type: 'UPDATE_BLOCKS_VALIDATION_ERRORS',
+							blockValidationErrorsByUid: blockValidationErrorsByUid
+						};
+					}
+				},
+				selectors: {
+					getBlockValidationErrors: function( state, uid ) {
+						return state.blockValidationErrorsByUid[ uid ] || [];
+					}
+				}
+			} );
+
+			wp.data.subscribe( function() {
+				var currentPost, thisValidationErrors;
+
+				// @todo Gutenberg currently is not persisting isDirty state if changes are made during save request. Block order mismatch.
+				// We can only align block validation errors with blocks in editor when in saved state.
+				if ( wp.data.select( 'core/editor' ).isEditedPostDirty() ) {
+					return;
+				}
+
+				currentPost = wp.data.select( 'core/editor' ).getCurrentPost();
+				thisValidationErrors = currentPost[ module.data.restValidationErrorsField ];
+				if ( thisValidationErrors && ! _.isEqual( lastValidationErrors, thisValidationErrors ) ) {
+					lastValidationErrors = thisValidationErrors;
+					module.updateBlocksValidationErrors(
+						currentPost.id,
+						wp.data.select( 'core/editor' ).getBlockOrder(),
+						thisValidationErrors
+					);
+
+					// @todo Create an notice in the editor when there are validation errors, beyond just for blocks.
+				}
+			} );
+		},
+
+		/**
+		 * Update blocks' validation errors in the store.
+		 *
+		 * @param {number}   postId           - Post ID.
+		 * @param {string[]} blockOrder       - Block order.
+		 * @param {Object[]} validationErrors - Validation errors.
+		 * @return {void}
+		 */
+		updateBlocksValidationErrors: function updateBlocksValidationErrors( postId, blockOrder, validationErrors ) {
+			var blockValidationErrorsByUid = {};
+
+			_.each( blockOrder, function( uid ) {
+				blockValidationErrorsByUid[ uid ] = [];
+			} );
+
+			_.each( validationErrors, function( validationError ) {
+				var i, source, matchedBlockUid;
+				if ( ! validationError.sources ) {
+					return;
+				}
+
+				// Find the inner-most nested block source only; ignore any nested blocks.
+				for ( i = validationError.sources.length - 1; 0 <= i; i-- ) {
+					source = validationError.sources[ i ];
+
+					if ( ! source.block_name || postId !== source.post_id ) {
+						continue;
+					}
+
+					// @todo Cross-check source.block_name with wp.data.select( 'core/editor' ).getBlock( matchedBlockUid ).name?
```

Comment:
Indeed, this is going to be required because my approach here to use the `blockOrder` along is somewhat naïve since a plugin could _add_ or _remove_ additional blocks during `the_content` filter. Nevertheless, this should hopefully be an edge case.